### ISSUE TITLE
eeshenggoh - User is not able to stake vesting more than once causing them to lose potential rewards

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-12-truflation-judging#158]
eeshenggoh

high

# User is not able to stake vesting more than once causing them to lose potential rewards

## Summary
The TrufVesting::stake() function enables users to stake their vested amount, predetermined by the Admin. The vesting period spans 1, 2, or 4 years, which is notably lengthy. As a consequence, claiming all tokens takes a considerable amount of time, resulting in users having the option to claim only a small amount within a short period, leading to limited voting power. The initial intention was to empower users with higher voting authority by allowing them to lock their entire vesting; however, the current implementation hinders users from doing so.
## Vulnerability Detail
The limitation of being able to call TrufVesting::stake() only once not only prevents users from staking and earning more rewards but also reduces their voting power. Additionally, if an admin seeks to increase the user's vesting using TrufVesting::setUserVesting() in the future, the user is unable to leverage the vested amount to stake and gain additional voting power.
## Impact
This restriction prevents users from staking their vested amounts, resulting in the inability to claim more rewards and a consequent reduction in their overall voting power.
## Code Snippet
https://github.com/sherlock-audit/2023-12-truflation/blob/main/truflation-contracts/src/token/TrufVesting.sol#L245-L247

To run `forge test --match-test testStakeTwice_Revert -vv`
```solidity
//TrufVesting.t.sol
    function testStakeTwice_Revert() external {
        console.log("PoC restake to veTRUF reverts");

        _setupVestingPlan();
        _setupExampleUserVestings();

        uint256 categoryId = 0;
        uint256 vestingId = 0;
        uint256 stakeAmount = 10e18;
        uint256 duration = 30 days;

        (uint256 amount,,, uint64 startTime) = vesting.userVestings(categoryId, vestingId, alice);

        uint256 trufBalanceBefore = trufToken.balanceOf(address(vesting));

        vm.startPrank(alice);

        vm.expectEmit(true, true, true, true, address(vesting));
        emit Staked(categoryId, vestingId, alice, stakeAmount, duration, 1);
        
        vesting.stake(categoryId, vestingId, stakeAmount, duration);

        assertEq(trufToken.balanceOf(address(veTRUF)), stakeAmount, "Staked amount is invalid");
        assertEq(trufToken.balanceOf(address(vesting)), trufBalanceBefore - stakeAmount, "Remaining balance is invalid");

        (uint128 lockupAmount,,,, bool lockupIsVesting) = veTRUF.lockups(alice, 0);

        assertEq(lockupAmount, stakeAmount, "Lockup amount is invalid");
        assertEq(lockupIsVesting, true, "Lockup vesting flag is invalid");

        _validateUserVesting(categoryId, vestingId, alice, amount, 0, stakeAmount, startTime);
        assertEq(vesting.lockupIds(categoryId, vestingId, alice), 1, "Lockup id is invalid");

        //Stake Again which reverts
        vm.expectRevert(abi.encodeWithSignature("LockExist()"));
        vesting.stake(categoryId, vestingId, stakeAmount, duration);
        vm.stopPrank();
    }
```
Result:
```solidity
[PASS] testStakeTwice_Revert() (gas: 1314893)
Logs:
  Stake vesting to veTRUF

Test result: ok. 1 passed; 0 failed; 0 skipped; finished in 7.96ms
Ran 1 test suites: 1 tests passed, 0 failed, 0 skipped (1 total tests)
```
## Tool used

Manual Review

## Recommendation
It is recommended to modify the implementation to allow users to stake multiple times for their vested amounts.
```diff
diff --git a/truflation-contracts/src/token/TrufVesting.sol b/truflation-contracts/src/token/TrufVesting.sol index 3c50d89..dbda239 100644
--- a/truflation-contracts/src/token/TrufVesting.sol
+++ b/truflation-contracts/src/token/TrufVesting.sol
@@ -242,9 +242,6 @@ contract TrufVesting is Ownable {
         if (amount == 0) {
             revert ZeroAmount();
         }
-        if (lockupIds[categoryId][vestingId][msg.sender] != 0) {
-            revert LockExist();
-        }
 
         UserVesting storage userVesting = userVestings[categoryId][vestingId][msg.sender];
 
```

Duplicate of #39
